### ISSUE TITLE
sap_hana/sap_swpm: Linux Special Characters

### ISSUE THREAD

**Issue by sean-freeman** [sap-linuxlab/community.sap_install#794]
We should probably also provide a warning in the `sap_hana_install` and `sap_swpm` Ansible Roles, regarding Linux Special Characters (aka. Metacharacters).

It is most noticeable when hdbuserstore executes successfully, but the subsequent login does not work. This is because the password has been incorrectly parsed by the Shell.

For more information, see [SAP Note 2667891 - R3trans gives authentication error after system copy and restart](https://me.sap.com/notes/2667891) which recommends escaping characters during data entry (i.e. hdbuserstore) or subsequently escaping characters for all logins.

A warning message during Ansible Role execution to state the password may cause issues, will be sufficient.

**Comment by berndfinger** [comment]
Is there a list of forbidden characters? SAP note [2667891](https://me.sap.com/notes/2667891) mentions `$` but only as an example.

Referenced SAP note [2250144](https://me.sap.com/notes/2250144) only refers to SAP note [3158257](https://me.sap.com/notes/3158257), which does not refer to valid passwords or special characters.

Referenced SAP note [3158257](https://me.sap.com/notes/3158257) does not refer to valid passwords or special characters either.

A first solution could be that we define a list or dict of forbidden special characters to be used for passwords and in the first step just use the character `$` as the only entry.

**Comment by sean-freeman** [comment]
I do not believe there is a specific list published by SAP, which is unfortunate. The below are the notes from the [Ansible Playbooks for SAP project FAQ page](https://github.com/sap-linuxlab/ansible.playbooks_for_sap/tree/main/docs#readme).


----


### SAP HANA password restrictions?

- Between 6 and 64 characters
- Alphanumerical, not advisable to use space character
- No restrictions on Special Characters

Reference:
- [SAP HANA Security Guide for SAP HANA Platform - Password Policy](https://help.sap.com/docs/SAP_HANA_PLATFORM/b3ee5778bc2e4a089d3299b82ec762a7/dce2826dbb571014be628d6279aeeaa3.html?locale=en-US)
- [SAP Note 2969917 - Can't use special characters like ! @ # $ % & in HANA user's password](https://me.sap.com/notes/2969917)


### SAP AnyDB password restrictions?

#### SAP Sybase ASE

No special recommendations

#### SAP MaxDB

Restricted to certain Special Characters `#$@_`. Must not begin with a digit.

#### IBM Db2

Avoid use of Special Character `$` which may cause automation errors when parsed

#### Oracle DB

Avoid use of Special Character `$` which may cause automation errors when parsed. Must not begin with a digit or underscore.


### SAP System / SAP NetWeaver password restrictions?

> Note: These are configurable in the Profile Parameters (`login/min_password_*` and `login/password_*`), below are default

- Between 3 and 40 characters
- Alphanumerical, not advisable to use space character
- Restricted to certain Special Characters `!"@$%&/()=?’*+~#-_.,;:{[]}\<>│`. Not advisible to use `\` or `"`

**Reference:**

For SAP NetWeaver Application Server (ABAP) see document [Password Rules - User and Role Administration - SAP NetWeaver Application Server for ABAP 7.52](https://help.sap.com/docs/SAP_NETWEAVER_AS_ABAP_752/c6e6d078ab99452db94ed7b3b7bbcccf/4ac3efb58c352470e10000000a42189c.html?locale=en-US).

For further information, please see [User Guides for System Provisioning with Software Provisioning Manager](https://help.sap.com/docs/SOFTWARE_PROVISIONING_MANAGER/30839dda13b2485889466316ce5b39e9/4901b2032db94b64bac87454dd94805b.html) which contains a list of different guides under two sections:

1. Installation Option of Software Provisioning Manager 2.0
    - Installation Guides - Application Server Systems - Software Provisioning Manager 2.0
2. Installation Option of Software Provisioning Manager 1.0
    - Installation Guides - Application Server Systems - Software Provisioning Manager 1.0

On each of these pages (for SWPM 1.0/2.0) there are documents (HTML/PDF) in a table with choice for `Database, Product Release, Operating System Platform, Technical Track`. Each document contains layered sections 'Planning > Basic Installation Parameters > SAP System Parameters' with Password limitation information.

For example, [SAP System Parameters - Installation of SAP ABAP Systems on UNIX : SAP HANA 2.0 - SWPM 2.0](https://help.sap.com/docs/SLTOOLSET/39c32e9783f6439e871410848f61544c/fe3f4554f82b1d5de10000000a44538d.html).

Please note, these guides are different than those listed on [Guide Finder for SAP NetWeaver and ABAP Platform](https://help.sap.com/docs/r/nwguidefinder).
